File: perflib/src/main/java/com/android/tools/perflib/heap/ClassInstance.java
/*
 * Copyright (C) 2008 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.tools.perflib.heap;

import com.android.annotations.NonNull;
import com.android.annotations.Nullable;
import com.android.annotations.VisibleForTesting;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;

public class ClassInstance extends Instance {

    private final long mValuesOffset;

    public ClassInstance(long id, @NonNull StackTrace stack, long valuesOffset) {
        super(id, stack);
        mValuesOffset = valuesOffset;
    }

    @VisibleForTesting
    @NonNull
    List<FieldValue> getFields(String name) {
        ArrayList<FieldValue> result = new ArrayList<FieldValue>();
        for (FieldValue value : getValues()) {
            if (value.getField().getName().equals(name)) {
                result.add(value);
            }
        }
        return result;
    }

    @NonNull
    public List<FieldValue> getValues() {
        ArrayList<FieldValue> result = new ArrayList<FieldValue>();

        ClassObj clazz = getClassObj();
        getBuffer().setPosition(mValuesOffset);
        while (clazz != null) {
            for (Field field : clazz.getFields()) {
                result.add(new FieldValue(field, readValue(field.getType())));
            }
            clazz = clazz.getSuperClassObj();
        }
        return result;
    }

    @Override
    public final void resolveReferences() {
        for (FieldValue fieldValue : getValues()) {
            if (fieldValue.getValue() instanceof Instance) {
                Instance referencedInstance = (Instance)fieldValue.getValue();
                referencedInstance.addReverseReference(fieldValue.getField(), this);
                if (getIsSoftReference() && fieldValue.getField().getName().equals("referent")) {
                    mSoftForwardReference = referencedInstance;
                } else {
                    mHardForwardReferences.add(referencedInstance);
                }
            }
        }
        mHardForwardReferences.trimToSize(); // Don't wait until the compactMemory stage to trim.
    }

    @Override
    public final void accept(@NonNull Visitor visitor) {
        visitor.visitClassInstance(this);
        for (Instance instance : mHardForwardReferences) {
            visitor.visitLater(this, instance);
        }
    }

    @Override
    public boolean getIsSoftReference() {
        return getClassObj().getIsSoftReference();
    }

    public final String toString() {
        return String
                .format("%s@%d (0x%x)", getClassObj().getClassName(), getUniqueId(), getUniqueId());
    }

    public boolean isStringInstance() {
        return getClassObj() != null && "java.lang.String".equals(getClassObj().getClassName());
    }

    @Nullable
    public final String getAsString() {
        return getAsString(Integer.MAX_VALUE);
    }

    @Nullable
    public final String getAsString(int maxDecodeStringLength) {
        int count = -1;
        int offset = 0;
        ArrayInstance charBufferArray = null;
        // In later versions of Android, the underlying storage format changed to byte buffers.
        ArrayInstance byteBufferArray = null;
        for (ClassInstance.FieldValue entry : getValues()) {
            if (charBufferArray == null && "value".equals(entry.getField().getName())) {
                if (entry.getValue() instanceof ArrayInstance) {
                    if (((ArrayInstance) entry.getValue()).getArrayType() == Type.CHAR) {
                        charBufferArray = (ArrayInstance) entry.getValue();
                    } else if (((ArrayInstance) entry.getValue()).getArrayType() == Type.BYTE) {
                        byteBufferArray = (ArrayInstance) entry.getValue();
                    }
                }
            } else if ("count".equals(entry.getField().getName())) {
                if (entry.getValue() instanceof Integer) {
                    count = (Integer) entry.getValue();
                }
            } else if ("offset".equals(entry.getField().getName())) {
                if (entry.getValue() instanceof Integer) {
                    offset = (Integer) entry.getValue();
                }
            }
        }

        if (byteBufferArray != null) {
            try {
                return new String(
                        byteBufferArray.asRawByteArray(
                                offset >= 0 ? offset : 0,
                                Math.max(Math.min(count, maxDecodeStringLength), 0)),
                        "UTF-8");
            } catch (UnsupportedEncodingException e) {
                return null;
            }
        }
        return charBufferArray == null
                ? null
                : new String(
                        charBufferArray.asCharArray(
                                offset >= 0 ? offset : 0,
                                Math.max(Math.min(count, maxDecodeStringLength), 0)));
    }

    public static class FieldValue {

        private Field mField;

        private Object mValue;

        public FieldValue(@NonNull Field field, @Nullable Object value) {
            this.mField = field;
            this.mValue = value;
        }

        public Field getField() {
            return mField;
        }

        public Object getValue() {
            return mValue;
        }
    }
}

File: perflib/src/main/java/com/android/tools/perflib/heap/ClassInstance.kt
/*
 * Copyright (C) 2008 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.perflib.heap

import com.android.annotations.VisibleForTesting
import java.io.UnsupportedEncodingException
import java.util.Locale
import kotlin.math.max
import kotlin.math.min

open class ClassInstance(id: Long, stack: StackTrace?, private val valuesOffset: Long)
    : Instance(id, stack) {
    open val values: List<FieldValue>
        get() = mutableListOf<FieldValue>().also { result ->
            buffer.setPosition(valuesOffset)
            tailrec fun collect(cl: ClassObj?) {
                if (cl != null) {
                    result.addAll(cl.fields.map { FieldValue(it, readValue(it.type)) })
                    collect(cl.superClassObj)
                }
            }
            collect(classObj)
        }

    override val isSoftReference: Boolean get() = classObj!!.isSoftReference
    val isStringInstance: Boolean get() = "java.lang.String" == classObj?.className
    val asString: String? get() = getAsString(Int.MAX_VALUE)

    @VisibleForTesting
    fun getFields(name: String): List<FieldValue> = values.filter { it.field.name == name }

    override fun resolveReferences() {
        for (fieldValue in values) {
            if (fieldValue.value is Instance) {
                fieldValue.value.addReverseReference(fieldValue.field, this)
                if (!isSoftReference || fieldValue.field.name != "referent") {
                    hardForwardReferences.add(fieldValue.value)
                }
            }
        }
        hardForwardReferences.trimToSize() // Don't wait until the compactMemory stage to trim.
    }

    override fun accept(visitor: Visitor) {
        visitor.visitClassInstance(this)
        for (instance in hardForwardReferences) {
            visitor.visitLater(this, instance)
        }
    }

    override fun toString() =
        String.format(Locale.US, "%s@%d (0x%x)", classObj!!.className, uniqueId, uniqueId)

    fun getAsString(maxDecodeStringLength: Int): String? {
        var count = -1
        var offset = 0
        var charBufferArray: ArrayInstance? = null
        // In later versions of Android, the underlying storage format changed to byte buffers.
        var byteBufferArray: ArrayInstance? = null
        for (entry in values) {
            when {
                charBufferArray == null && "value" == entry.field.name ->
                    if (entry.value is ArrayInstance) {
                        when (entry.value.arrayType) {
                            Type.CHAR -> charBufferArray = entry.value
                            Type.BYTE -> byteBufferArray = entry.value
                        }
                    }
                "count" == entry.field.name && entry.value is Int -> count = entry.value
                "offset" == entry.field.name && entry.value is Int -> offset = entry.value
            }
        }
        fun<T> mkRaw(mk: (Int, Int) -> T) = mk(max(offset, 0),
                                               max(min(count, maxDecodeStringLength), 0))
        return when {
            byteBufferArray != null ->
                try {
                    String(mkRaw(byteBufferArray::asRawByteArray), Charsets.UTF_8)
                } catch (e: UnsupportedEncodingException) {
                    null
                }
            charBufferArray != null -> String(mkRaw(charBufferArray::asCharArray))
            else -> null
        }
    }

    class FieldValue(val field: Field, val value: Any?)
}
